examples/general/basic_dead_store/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+    expr: &Expr,
+    arr_string: &String,
+    tcx: &LateContext<'_>,
+) -> Option<(String, u128, Span)> {
+    let index_expr = get_parent_expr(tcx, expr)?;
+    if let ExprKind::Index(array, index, _span) = index_expr.kind {
+        if array.hir_id == expr.hir_id {
+            let assign_expr = get_parent_expr(tcx, index_expr)?;
+            if let ExprKind::Assign(target, _value, assignment_span) = assign_expr.kind {
+                if target.hir_id == index_expr.hir_id {
+                    if let ExprKind::Lit(lit) = index.kind {
+                        if let LitKind::Int(index, _type) = lit.node {
+                            return Some((arr_string.to_string(), index, assignment_span));
+                        }
+                    }
+                }
+            }
+        }
+    }
+    None
+}
+
+impl<'tcx> LateLintPass<'tcx> for BasicDeadStore {
+    // Given an Expression:
+    //  - If we are looking at an array being indexed by a literal:
+    //      - if we have seen this array being indexed at this literal before
+    //          - then we found a dead store
+    //      - otherwise, we save this instance in a vec V
+    //  - Otherwise, clear all stored instances of this expression in the vec V
+    fn check_expr(
+        &mut self,
+        ctx: &rustc_lint::LateContext<'tcx>,
+        expr: &'tcx rustc_hir::Expr<'tcx>,
+    ) {
+        if let ExprKind::Path(ref qpath) = expr.kind {
+            let array_resolution = ctx.qpath_res(qpath, expr.hir_id);
+            let arr_string = format!("{array_resolution:?}");
+
+            if let Some((array_string, v, span)) =
+                is_assignment_to_array_indexed_by_literal(expr, &arr_string, ctx)
+            {
+                let in_common = self.get_pairs_with_same_name_idx(&array_string, &v);
+                if in_common.is_empty() {
+                    // If there are no saved instances of this array being assigned indexed at the
+                    // same index, save this instance
+                    self.arr_and_idx_vec.push((array_string, v, span));
+                } else {
+                    // Otherwise, we are storing again into the same index
+                    // unwrap: `in_common` is guaranteed to have at least one element
+                    span_lint_and_help(
+                        ctx,
+                        BASIC_DEAD_STORE,
+                        span,
+                        "reassigning the same array position without using it",
+                        Some(in_common.first().unwrap().2),
+                        "original assignment was here",
+                    );
+                }
+            } else {
+                // If we are using the array in a way that is not an assignment to a certain
+                // position, then it is being used. Therefore, we need to clear all
+                // stored instances of this array
+                self.clear_stores_of(&arr_string);
```

Comment:
No action is needed, but this gives me pause, because it makes assumptions about the order in which expressions are processed.

I suspect those assumptions hold in practice, but it would be nice if we had a solution that didn't rely on them.

Reply:
I agree -- initially, I was checking on functions and iterating over the statements of the basic blocks but I was having issues finding all nested expressions within.